File: lang/ruby/test/fun/vac/data_structures/ArrayListTest.rb
require "fun/vac/data_structures/ArrayList"
require "fun/vac/util/TestRunner"

def testArrayList()
    size = 8192

    list = ArrayList::ArrayList.new(0)

    for i in 1 .. size
        list.append(i)
    end

    list.shrink()

    if list.size() != size
        return false
    end

    if list.capacity() != size
        return false
    end

    for i in 0 ... size
        if list.get(i) != i + 1
            return false
        end
    end

    for i in 0 ... size
        list.set(i, size - i)
    end

    for i in 0 ... size
        if list.get(i) != size - i
            return false
        end
    end

    x, y = 0, 0

    i, j = 0, size - 1
    while i < j
        x = list.get(i)
        y = list.get(j)

        list.remove(i)
        list.insert(i, y)
        list.remove(j)
        list.insert(j, x)
        i, j = i + 1, j - 1
    end

    for i in 0 ... size
        if list.get(i) != i + 1
            return false
        end
    end

    i = size
    while i >= 1
        if list.back() != i
            return false
        end
        list.eject()
        i -= 1
    end

    list.shrink()

    unless list.isEmpty()
        return false
    end

    if list.capacity() != 0
        return false
    end

    return true
end

if __FILE__ == $0
    TestRunner.parseTest(testArrayList())
end

File: lang/nodejs/test/fun/vac/data_structures/ArrayListTest.js
"use strict"

const ArrayList  = require("fun/vac/data_structures/ArrayList")
const TestRunner = require("fun/vac/util/TestRunner")

const testArrayList = () => {
    let size = 8192

    const list = new ArrayList.ArrayList(0)

    for (let i = 1; i <= size; i++) {
        list.append(i)
    }

    list.shrink()

    if (list.size() !== size) {
        return false
    }

    if (list.capacity() !== size) {
        return false
    }

    for (let i = 0; i < size; i++) {
        if (list.get(i) !== i + 1) {
            return false
        }
    }

    for (let i = 0; i < size; i++) {
        list.set(i, size - i)
    }

    for (let i = 0; i < size; i++) {
        if (list.get(i) !== size - i) {
            return false
        }
    }

    let x = 0
    let y = 0

    for (let i = 0, j = size - 1; i < j; i++, j--) {
        x = list.get(i)
        y = list.get(j)

        list.remove(i)
        list.insert(i, y)
        list.remove(j)
        list.insert(j, x)
    }

    for (let i = 0; i < size; i++) {
        if (list.get(i) !== i + 1) {
            return false
        }
    }

    for (let i = size; i >= 1; i--) {
        if (list.back() !== i) {
            return false
        }
        list.eject()
    }

    list.shrink()

    if (!list.isEmpty()) {
        return false
    }

    return list.capacity() === 0
}

if (module === require.main) {
    TestRunner.parseTest(testArrayList())
}
